proposals/0000-local-modules.rst

```diff
@@ -783,16 +783,18 @@ component of the overall proposal. Later pieces can be chosen piecemeal.
       modules influence scoping only, but not type-checking or dependency
       (which remain constrained by compilation units, as they are today).
 
-   #. Let the ambient QEnv of the local module (call it ``M``) be ``E``.
-      The export QEnv
-      of ``M`` is the result of interpreting ``exports`` (if given)
-      with respect to ``E``.
-
-      If ``exports`` is not given, then the export environment of ``M``
-      includes a mapping from occurrence name to original name
-      for each definition within the local module. For each nested local
-      module ``N`` within ``M`` with export QEnv ``N_E``, the
-      export QEnv of ``M`` includes ``qualify(N, N_E)``.
+   #. Let the ambient QEnv of the local module (call it ``M``, located within
+      top-level module ``Top`` and under other local modules ``modids``) be ``E``.
+
+      If ``exports`` is given, the export QEnv
+      of ``M`` is the result of interpreting ``exports`` with respect to ``E``.
+
+      If ``exports`` is not given, then the export QEnv ``X`` of ``M``
+      is the subset of ``E`` containing all mappings to original names that
+      appear within ``M``. That is, ::
+
+        X = { qual ↦ orig | qual ↦ orig ∈ E
+                          , orig = (Top, modids.M._, _occ) }
```

Comment:
Hmm, I think this implies the following behavior. Intended?
```
module Top
module qualified M where foo = True
module qualified B where import module qualified M as C
_ = M.B.C.foo 
```
If I read the spec correctly, this would work: The ambient QEnv of `M` contains the ambient QEnv of `TOP` which has an entry `B.C.foo ↦  (Top, M, foo)`, so it passes the filter, thus `B.C.foo` is part of the exports, and thus `M.B.C.foo` comes into scope.

If this is not intended, then you might have to distinguish between the “inherited ambient QEnv” of a scope (stuff coming from declarations in surrounding scopes) and the “intrinsic QEnv” (stuff coming from declarations, including `import`, in the current scope). Then the `exports` of a local module would be filtered based on that. Not sure.

Reply:
No, you are right, I made assumptions about the proposal that are not true. Let me try again…

Ok, `import module M` only ever _strips_ `M` from stuff that’s already in scope, but no renaming is possible…

How about this, does this work?
```
module Top
module qualified M where foo = True
module qualified B(foo) where import module M (foo)
_ = M.B.foo 
```
If I (now hopefully) read the spec correctly, the ambient QEnv of M contains the ambient QEnv of `Top` which has an entry `B.foo ↦ (Top, M, foo)`, so it passes the filter, thus `B.foo` is part of the exports, and thus `M.B.C.foo` comes into scope?

Reply:
No, I don't think so. The proposal says

```

        X = { qual ↦ orig | qual ↦ orig ∈ E
                          , orig = (Top, modids.M._, _occ) }
```

where `E` is the ambient QEnv in the local module `M`, and `modids` (in your example) is empty. Thus, `X` will be `foo ↦ (Top, M, foo)`. Thus the ambient QEnv outside of any local module will be

```
{ M.foo ↦ (Top, M, foo), B.foo ↦ (Top, M, foo), Top.M.foo ↦ (Top, M, foo), Top.B.foo ↦ (Top, M, foo) }
```

and that's it.

Reply:
Sorry for making you debug my understanding of the proposal. What is the value of `E`, the ambient QEnv within `M`, in this example?

Reply:
You have laid a trap! Well done.

(More explicitly: I believe @nomeata is drawing attention to the fact that, according to the definitions above, the ambient QEnv in `M` will include bindings like `Top.M.foo ↦ (Top, M, foo)`, which would end up in `X`. Yet I don't want `M` to export `Top.M.foo`, which would mean that `M.Top.M.foo` is in scope.)

The best way I can see fixing this problem is to maintain a stack of ambient QEnvs, where entering a local module adds one to the stack, and name lookup proceeds linearly through the stack. I actually don't think this is such a bad way of thinking of things, but it does add an extra layer of complexity. Do you see an alternative?

Reply:
No, I think it's unavoidable. Not sure if “ambient” is still a good adjective, though, then, it sounds (to me) as if it's already the union.

You might still additionally have to keep “defined here” and “imported” apart to get the semantics of the default export list right, which should not be affected by imports.  Matching on the original name after throwing the two together might not work well.